org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -905,6 +916,58 @@ public boolean getWantClientAuth() {
         return want_client_auth;
     }
 
+    /**
+     * Set a specific list of protocols to negotiate next for ALPN support.
+     */
+    public void setApplicationProtocols(String[] protocols) {
+        alpn_protocols = protocols;
+    }
+
+    /**
+     * Get the most recently negotiated application protocol.
+     *
+     * Note that NSS only allows selection on the initial handshake so
+     * this is implemented via a call to getHandshakeApplicationProtocol().
+     */
+    public String getApplicationProtocol() {
+        return getHandshakeApplicationProtocol();
+    }
+
+    /**
+     * Get the application protocol negotiated during the initial handshake.
+     */
+    public String getHandshakeApplicationProtocol() {
+        if (session == null) {
+            return null;
+        }
+
+        return session.getNextProtocol();
+    }
+
+    /**
+     * Helper method for implementations: encodes ALPN protocols into wire
+     * format (8-bit length prefixed byte encoding).
+     */
+    public byte[] getALPNWireData() {
+        int length = 0;
+        for (String protocol : alpn_protocols) {
+            length += 1 + protocol.getBytes().length;
```

Comment:
Do we need to use UTF-8 specifically? Otherwise `getBytes()` will use the system's default encoding. Or if the protocol names are ASCII anyway then it doesn't matter.

Reply:
Officially an ALPN protocol ID is a sequence of bytes and they are registered with IANA as such; see https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids.  Therefore I recommend the JSS interface deal with protocol IDs as `byte[]` the whole way through, and defer conversion to/from other types (such as `String`) to the library user.  This is what I did in my other ALPN PR #640.

Reply:
This isn't actually possible. This PR is against the JCA-standard `javax.net.ssl` implementation that I've added to JSS. It resides under the `org.mozilla.jss.ssl.javax` package namespace to avoid confusing it with the custom, legacy "SSLSocket" implementation under `org.mozilla.jss.ssl`. 

In particular, looking at the interface we're given:

 - We get the output as a `String`: [`SSLSocket.getApplicationProtocol()`](https://docs.oracle.com/javase/9/docs/api/javax/net/ssl/SSLSocket.html#getApplicationProtocol--), [`SSLEngine.getApplicationProtocol()`](https://docs.oracle.com/javase/9/docs/api/javax/net/ssl/SSLEngine.html#getApplicationProtocol--)
 - We get the "input"/protocol choices as a `String`: [`SSLEngine.setHandshakeApplicationProtocolSelector(...)`](https://docs.oracle.com/javase/9/docs/api/javax/net/ssl/SSLEngine.html#setHandshakeApplicationProtocolSelector-java.util.function.BiFunction-)

So we either choose our own encoding and force it throughout, or let the user control it via system default.

It looks like the JDK11 always forces UTF-8:
https://github.com/openjdk/jdk11u-dev/blob/master/src/java.base/share/classes/sun/security/ssl/AlpnExtension.java#L100

So perhaps this will be good enough for us.

Reply:
It will be good enough for us.  But it is seriously flawed w.r.t. how the ALPN extension and protocol ID is actually specified.  Java's fault, we just have to put up with it.

Reply:
I mean, I don't disagree:

```java
        byte[] a = new byte[]{ (byte) 0xFA, (byte) 0xFA };
        String b = new String(a, StandardCharsets.UTF_8);
        byte[] c = b.getBytes(StandardCharsets.UTF_8);
        System.out.println(a.length + " " + a[0] + " " +a[1]);
        System.out.println(b); 
        System.out.println(c.length + " " + c[0] + " " + c[1]);
```

But also.. If you and the server both negotiate GREASE'd values... What does that even mean? What protocol are you actually negotiating there? It really isn't well defined. So they just hope the rest of the (valid) IANA protocol values are otherwise valid.

Reply:
(I say negotiate because NSS and the protocol doesn't give the client a way to see what the server has for ALPN data priorities, and for that matter, the String comparison all happens on the server-side anyways, so we should probably coerce byte -> String and use `String.equals(...)`, rather than trying to go from String -> Byte and compare byte arrays... That ensure we probably have the same encoded representation rather than dealing with differences there...)

Reply:
Well, there's nothing we can do about it but to descend slightly further into this rabbit hole:

1. You don't negotiate GREASE (servers MUST ignore GREASE values).  But if you want to *send* GREASE in your ClientHello, 15 of the 16 defined values are impossible to send.
2. And if any other ALPN values get defined that are not the UTF-8 encoding of some string, then you can't send those either.

On your second comment, no, the string comparison is essential on the client side too: the client will usually need to check what protocol was negotiated.

As for how to compare, given the interface imposed on us I have no objection to `String.equals()`.

Reply:
I suppose... This is why we have JSSEngine and JSSParameters after all... We could add a custom getter/setter which returns the raw byte values... Thoughts?

Reply:
@frasertweedale -- I've brought this up for discussion on the OpenJDK `security-dev` mailing list and CC'd you. That was the approach suggested by `#java`.

Reply:
Yes, we could implement our own methods that deal with `byte[]`, if the requirement arises.

I don't think the need is there right now (I do not need to send grease or handle a non-UTF-8 protocol ID).

I'm just annoyed at such a mistake in a foundational API.  So easliy avoided, if they had just read the spec and implemented what is says.